test/integration/apiserver/discovery/discovery_test.go

```diff
@@ -0,0 +1,224 @@
+
+	discoveryendpoint "k8s.io/apiserver/pkg/endpoints/discovery/aggregated"
+	genericfeatures "k8s.io/apiserver/pkg/features"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	k8sscheme "k8s.io/client-go/kubernetes/scheme"
+
+	"k8s.io/kubernetes/test/integration/framework"
+)
+
+type kubeClientSet = kubernetes.Interface
+type aggegatorClientSet = aggregator.Interface
+type testClientSet struct {
+	kubeClientSet
+	aggegatorClientSet
+}
+
+func (t testClientSet) Discovery() discovery.DiscoveryInterface {
+	return t.kubeClientSet.Discovery()
+}
+
+var _ kubernetes.Interface = &testClientSet{}
+var _ aggregator.Interface = &testClientSet{}
+
+// Spins up an api server which is cleaned up at the end up the test
+// Returns some kubernetes clients
+func setup(t *testing.T) (context.Context, testClientSet, context.CancelFunc) {
+	ctx, cancelCtx := context.WithCancel(context.Background())
+
+	server := kubeapiservertesting.StartTestServerOrDie(t, nil, nil, framework.SharedEtcd())
+	t.Cleanup(server.TearDownFn)
+
+	client := testClientSet{
+		kubeClientSet:      kubernetes.NewForConfigOrDie(server.ClientConfig),
+		aggegatorClientSet: aggregator.NewForConfigOrDie(server.ClientConfig),
+	}
+	return ctx, client, cancelCtx
+}
+
+func registerAPIService(ctx context.Context, client aggregator.Interface, gv metav1.GroupVersion, service FakeService) error {
+	port := service.Port()
+	if port == nil {
+		return errors.New("service not yet started")
+	}
+	// Register the APIService
+	patch := v1.APIService{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: gv.Version + "." + gv.Group,
+		},
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "APIService",
+			APIVersion: "apiregistration.k8s.io/v1",
+		},
+		Spec: v1.APIServiceSpec{
+			Group:                 gv.Group,
+			Version:               gv.Version,
+			InsecureSkipTLSVerify: true,
+			GroupPriorityMinimum:  1000,
+			VersionPriority:       15,
+			Service: &v1.ServiceReference{
+				Namespace: "default",
+				Name:      service.Name(),
+				Port:      port,
+			},
+		},
+	}
+
+	_, err := client.
+		ApiregistrationV1().
+		APIServices().
+		Create(context.TODO(), &patch, metav1.CreateOptions{FieldManager: "test-manager"})
+	return err
+}
+
+func unregisterAPIService(ctx context.Context, client aggregator.Interface, gv metav1.GroupVersion) error {
+	return client.ApiregistrationV1().APIServices().Delete(ctx, gv.Version+"."+gv.Group, metav1.DeleteOptions{})
+}
+
+var scheme = runtime.NewScheme()
+var codecs = runtimeserializer.NewCodecFactory(scheme)
+var serialize runtime.NegotiatedSerializer
+
+func init() {
+	// Add all builtin types to scheme
+	k8sscheme.AddToScheme(scheme)
+	info, ok := runtime.SerializerInfoForMediaType(codecs.SupportedMediaTypes(), runtime.ContentTypeJSON)
+	if !ok {
+		panic("failed to create serializer info")
+	}
+
+	serialize = runtime.NewSimpleNegotiatedSerializer(info)
+}
+
+func TestAggregatedAPIServiceDiscovery(t *testing.T) {
```

Comment:
Where is the corresponding test for CRDs and for overlapping CRDs and apiservices.

Reply:
Added